Add documentation about the XR action map (#6633)
godotengine/godot-docs: tutorials/xr/xr_action_map.rst

For your first XR game/application we highly recommend starting with just a single action set and not over-engineer things.

For our walk through in this document we will thus create a single action set called ``my_first_action_set``, we do this by pressing the ``Add action set`` button:

.. image:: img/xr_my_first_action_set.webp

The columns in our table are as follows:

.. list-table::
  :class: wrap-normal
  :width: 100%
  :widths: 7 23 70
  :header-rows: 1

  * - Col
    - Value
    - Description
  * - 1
    - my_first_action_set
    - This is the internal name of the action set. OpenXR doesn't specify specific restrictions on this name other then size, however some XR runtimes will not like spaces or special characters.
  * - 2
    - My first action set
    - This is a human-readable name for the action set. Some XR runtimes will display this name to the end user, for example in configuration dialogs.
  * - 3
    - 0

>     - This is the priority of the action set. If multiple active action sets have actions bound to the same controllers inputs or outputs, the action set with the highest priority value will determine the action that is updated.

Reviewer: What occurs when multiple action sets have the same priority? 
Might be worth clarifying as a `Note`.

Author: I think that is undefined, same as having multiple actions in one action set bound to the same input.

Author: Ok, I've added a note to the bottom of this section giving some clarification here. I underline as with my discussion with @rsjtdrjgfuzkfg that I'm unsure of this as its not well defined in the specification. I've asked questions but not yet received suitable answers, I also get the feeling this may be in flux.

Dirk, I also added a small remark to this note about the haptic bindings. Again we simply don't know for sure what is allowed and I wouldn't be surprised if different runtimes do different things here.

Actions
=======

In the XR action map, actions are the entities that your game/application will interact with. For instance, we can define an action ``Shoot`` and the input bound to that action will trigger the ``button_pressed`` signal on the relevant :ref:`XRController3D <class_xrcontroller3d>` node in your scene with ``Shoot`` as the ``name`` parameter of the signal.

You can also poll the current state of an action. :ref:`XRController3D <class_xrcontroller3d>` for instance has an ``is_button_pressed`` method.

> Actions can be used for both in- and output. Each action has a type that defines the type of in- or output it can operate on.

Reviewer: Should be better to fully write `input` instead of using `in-`

Author: yes agree, much clearer

> The type ``Bool`` is used for buttons while type ``Float`` is used for triggers and other analogue inputs. These two are special as they are the only ones that are interchangeable. OpenXR will handle conversions between ``Bool`` and ``Float`` inputs and actions. You can get the value of a ``Float`` type action by calling the method ``get_value`` on your :ref:`XRController3D <class_xrcontroller3d>` node, and it emits the ``input_value_changed`` signal when changed.

Reviewer: Instead of `is used for buttons`, we could state `is used for discrete input like buttons`.

Similarly, `is used for triggers and other analogue inputs` can be replaced with `is used for analogue inputs like triggers`.

Author: yeah, flows better, thanks:)

Lets complete our list of actions for a very simple shooting game/application:

.. image:: img/xr_all_actions.webp

The actions we have added are:

  * movement, which allows the user to move around outside of normal room scale tracking.
  * grab, which detects the user wants to hold something.
  * shoot, which detects the user wants to fire the weapon they are holding.
  * haptic, which allows us to output haptic feedback.

Now note that we don't make a difference between the left and right hand. This is something that is determined at the next stage. We've implemented the action system in such a way that you can bind the same action to both hands, and the correct :ref:`XRController3D <class_xrcontroller3d>` node will emit the signal.

.. warning::
  Note that for both grab and shoot we've used the ``Bool`` type. As mentioned before, OpenXR does automatic conversions from an analogue controls however not all XR Runtimes currently apply sensible thresholds.

  We recommend as a workaround to use the ``Float`` type when interacting with triggers and grip buttons and apply your own threshold.

>   For A/B/X/Y buttons the ``Bool`` type works fine.

Reviewer: This may be confusing since not all controllers have the same button layouts. I'm not sure if it adds to the `warning` so it could be removed.

Author: I've changed the text here, I also added a note giving some more clarity around click and touch variants of inputs.

  For example, when the Touch controller profile is used any of the following scenarios could be true:

    * we could be using a Quest 1 controller,
    * we could be using a Quest 2 controller,
    * we could be using a Quest Pro controller but no Quest Pro profile was given or the XR runtime being used does not support the Quest Pro controller,
    * it could be a completely different controller for which no profile was given but the XR runtime is using the touch bindings as a base.

  Ergo, there currently is no way to know with certainty, which controller the user is actually using.

.. warning::
  Finally, and this trips up a lot of people, the bindings aren't set in stone. It is fully allowed, and even expected, that an XR runtime allows a user to customise the bindings.

  At the moment none of the XR runtimes offer this functionality though SteamVR has an existing UI from OpenVRs action map system that is still accessible. This is actively being worked on however.

Our first controller binding
----------------------------

> Lets setup our first controller binding and we'll do so for the Touch controller.

Reviewer: `Let's setup`

The XR action map system exposes input, positional data and output for XR controllers to your game. It does this by exposing named actions that can be tailored to your game and binding these to the actual inputs and outputs on your XR devices.

As the XR action map is currently part of the OpenXR module, OpenXR needs to be enabled in your project settings to expose it:

.. image:: img/openxr_settings.png

You will then find the XR Action Map interface in the bottom of the screen:

.. image:: img/xr_action_map.webp

.. note::
	Godots build in input system has many things in common with the XR action map system. In fact our original idea was to add functionality to the existing input system and expose the data to the OpenXR action map system. We may revisit that idea at some point but as it turns out there were just too many problems to overcome. To name a few:

> 	 * Godots input system mainly centers around button inputs, XR adds triggers, axis, poses and haptics (output) into the mix. This would greatly complicate the input system with features that won't work for normal controllers or contrast with the current approach. It was felt this would lead to confusion for the majority of Godot users.

Reviewer 2: ```suggestion
    * Godot's input system mainly centers around button inputs, XR adds triggers, axis, poses and haptics (output) into the mix. This would greatly complicate the input system with features that won't work for normal controllers or contrast with the current approach. It was felt this would lead to confusion for the majority of Godot users.
```

	 * Godots input system works with raw input data that is parsed and triggers emitting actions, this input data is made available to the end user. OpenXR completely hides raw data and does all the parsing for us, we only get access to already parsed action data. This inconsistency is likely to lead to bugs when an unsuspecting user tries to use an XR device as a normal input device.
	 * Godots input system allows changes to what inputs are bound to actions in runtime, OpenXR does not.
   * Godots input system is based on device ids which are meaningless in OpenXR.

	This does mean that a game that mixes traditional inputs with XR controllers will have a separation. For most applications either one or the other is used and this is not seen as a problem. In the end, it's a limitation of the system.

The default action map
======================

Godot will automatically create a default action map if no action map file is found.

.. warning::

> 	This default map was designed to help developers port their XR games from Godot 3 to Godot 4. As a result this map essentially binds all known inputs on all controllers supported by default, to actions one on one. This is not a good example of setting up an action map but it does allow a new developer to have a starting point when they want to become familiar with Godot XR without designing a proper action map for their game first.

Author: No, just that I'm writing this on the assumption that we may not change the default map in time and even if we change it, it will likely still be a default map to give people a quick start only.

I'm still very keen on continueing that discussion, we may want to pick it up in next weeks XR meeting.

For this walk through we're going to start with a blank action map. You can simply delete the "Godot action set" entry at the top by pressing the trash can icon and it will clear out all actions. You can also remove the controllers that you do not wish to setup, more on this later.

Action sets
===========

.. note::
  Before we dive in, you will see the term XR runtime used throughout this document. With XR runtime we mean the software that is controlling and interacting with the AR or VR headset and exposed to us through an API such as OpenXR. So for Steam this is SteamVR, for Meta on desktop this is the Oculus Client (including when using Quest link), for Meta on Quest this is the Quests native OpenXR client, on Linux this could be Monado, etc.

> The action map allows us to organise our actions in sets. These sets can be individually enabled/disabled.

Author: No, I always make those mistakes so thanks for pointing them out. I don't know if it is because I live in Oz or whether I learned English in Holland but I just can't seem to shake defaulting to British spelling.

The concept here is that you could have different sets that provide bindings in different scenarios. You could have a set ``Character control`` for when you're walking around, ``Vehicle control`` for when you're operating a vehicle, ``Menu`` for when a menu is open, etc.
Only the action set applicable to the current state of your game can then be enabled.

This is especially important if you wish to bind the same input on a controller to a different action. For instance:

 * in your ``Character control`` set you may have an action ``Jump``, 
 * in your ``Vehicle control`` set you may have an action ``Accelerate``, 
 * in your ``Menu`` set you may have an action ``Select``.

All are bound to the trigger on your controller.

OpenXR will only bind an input or output to a single action. If the same input or output is bound to multiple actions the one in the active action set with the highest priority will be the one updated/used. So in our above example it will thus be important that only one action set is active.

For your first XR game we highly recommend starting with just a single action set and not over engineer things.

> For our walk through in this document we will thus create a single action set called ``my_first_action_set``, we do this by pressing the ``Add action set`` button:

Author: I do make an example like that in the paragraph (or two) above but I do agree that "my first action set" is a bit of a cheeky wink to the many "my first ..." tutorials, the screenshots are pesky to redo but it might be worth the change, not sure.

.. image:: img/xr_my_first_action_set.webp

The columns in our table are as follows:

.. list-table::
  :class: wrap-normal
  :width: 100%
  :widths: 7 23 70
  :header-rows: 1

  * - Col
    - Value
    - Description
  * - 1
    - my_first_action_set
    - This is the internal name of the action set. OpenXR doesn't specify specific restrictions on this name other then size, however some XR runtimes will not like spaces or special characters.
  * - 2
    - My first action set

>     - This is the description of the action set presented to the end user.

Author: To be honest, I have no idea. The OpenXR spec seems to suggest localization is applied, but I have not looked into the mechanism of this.

Actions can also be used to poll that current state. :ref:`XRController3D <class_xrcontroller3d>` for instance has an ``is_button_pressed`` method.

Each action has a type that defines the type of state it manages. Inputs and outputs on controllers are equally typed.

The type ``Bool`` is used for buttons while type ``Float`` is used for triggers and other analogue inputs. These two are special as they are the only ones that are interchangeable. OpenXR will handle conversions between ``Bool`` and ``Float`` inputs and actions. You can get the value of a ``Float`` type action by calling the method ``get_value`` on your :ref:`XRController3D <class_xrcontroller3d>` node, and it emits the ``input_value_changed`` signal when changed.

.. note::
	Where analogue inputs are queried as buttons a threshold is applied. This threshold is currently managed by the XR runtime. There are settings on some runtimes that allow us control over these thresholds. There are plans to further extend this capability however currently development around this is still ongoing in Godot.

> The type ``Vector2`` defines the input as an axis input. Touchpads, thumbsticks and similar inputs are exposed as vectors. You can get the value of a ``Vector2`` type action by calling the method ``get_axis`` on your :ref:`XRController3D <class_xrcontroller3d>` node, and it emits the ``input_axis_changed`` signal when changed.

Author: The type names come from the OpenXR spec, while the method names are a left over from how things were in Godot 3.
I didn't realize this discrepancy until I wrote this document. We really should change the method and signal names but I wonder if it's too close to Godot 4 release to do that.

> The type ``Vector2`` defines the input as an axis input. Touchpads, thumbsticks and similar inputs are exposed as vectors. You can get the value of a ``Vector2`` type action by calling the method ``get_axis`` on your :ref:`XRController3D <class_xrcontroller3d>` node, and it emits the ``input_axis_changed`` signal when changed.

Reviewer 3: Imho *after* the Godot 4 release is a worse idea than doing it now, so I'd prefer to change it now if I'm not the only one finding it weird. The real question is whether people with authority can agree on the same naming schema in time :D

(I see two obvious solutions: Simply replace `axis` -> `vector2`, `button` -> `bool` and `value` -> `float`, or go semantic like `axis` -> `2d`, `button` -> `toggle`, `value` -> `1d` or maybe `float`. I assume the first one is easier to form consensus around as it uses an existing standard as basis.)

> The type ``Vector2`` defines the input as an axis input. Touchpads, thumbsticks and similar inputs are exposed as vectors. You can get the value of a ``Vector2`` type action by calling the method ``get_axis`` on your :ref:`XRController3D <class_xrcontroller3d>` node, and it emits the ``input_axis_changed`` signal when changed.

Author: I'm in favor of sticking to the names OpenXR handles (so bool, float, vector2) as those are also the variable types in Godot.

> The type ``Vector2`` defines the input as an axis input. Touchpads, thumbsticks and similar inputs are exposed as vectors. You can get the value of a ``Vector2`` type action by calling the method ``get_axis`` on your :ref:`XRController3D <class_xrcontroller3d>` node, and it emits the ``input_axis_changed`` signal when changed.

Author: This is now implemented in https://github.com/godotengine/godot/pull/71830
I will change the documentation once that is merged (in a separate PR if we merge the documentation as is)

> The type ``Vector2`` defines the input as an axis input. Touchpads, thumbsticks and similar inputs are exposed as vectors. You can get the value of a ``Vector2`` type action by calling the method ``get_axis`` on your :ref:`XRController3D <class_xrcontroller3d>` node, and it emits the ``input_axis_changed`` signal when changed.

Author: Done

  * - 1
    - aim_pose
    - This is the internal name of the action. OpenXR doesn't specify specific restrictions on this name other then size, however some XR runtimes will not like spaces or special characters.
  * - 2
    - Aim pose
    - This is the description of the action set presented to the end user.
  * - 3
    - Pose
    - The type of this action.

OpenXR defines a number of input poses that are commonly available for controllers and can thus be bound however there is no rule which poses are supported for different controllers. The poses OpenXR currently defines are:

 * The aim pose on most controllers is positioned slightly infront of the controller and aims forward. This is a great pose to use for laser pointers or to align the muzzle of a weapon with.
 * The grip pose on most controllers is positioned where the grip button is placed on the controller. The orientation of this pose differs between controllers and can differ for the same controller on different XR runtimes.
 * The palm pose on most controllers is positioned in the center of the palm of the hand holding the controller. This is a new pose that is not available on all XR runtimes.

 .. note::

>   If hand tracking is used there are currently big differences in implementations between the different XR runtimes. As a result the action map is currently not suitable for hand tracking. Work is being done on this so stay tuned.

Author: The skeleton pose is only one element of it, there is a lot more in the oven around this. I'll be covering more of it when I do the help page on hand tracking.

In this case I'm referring to hand gestures triggering actions. Each platform does different things here and none currently use the action map but again, things are in the oven so its likely to improve. We'll deal with it once OpenXR has proper support for this.

Lets complete our list of actions for a very simple shooting game:

.. image:: img/xr_all_actions.webp

The actions we have added are:

 * movement, which allows the user to move around outside of normal room scale tracking.
 * grab, which detects the user wants to hold something.
 * shoot, which detects the user wants to fire the weapon they are holding.

>  * haptic, which allows us to output haptic feedback.

Author: From how I read the spec at this moment I don't think that's allowed even though it does make sense to me to do what you suggest. I might raise that point in the WG.

>  * haptic, which allows us to output haptic feedback.

Author: So far not much conclusive feedback on this, I think we should keep things the way they are in the documentation for now and expand on it when we learn more

Now note that we don't make a difference between the left and right hand. This is something that is determined at the next stage. We've implemented the action system in such a way that you can bind the same action to both hands, and the correct :ref:`XRController3D <class_xrcontroller3d>` node will emit the signal.

.. warning::
  Note that for both grab and shoot we've used the ``Bool`` type. As mentioned before, OpenXR does automatic conversions from an analogue controls however not all XR Runtimes currently apply sensible thresholds.

>   We recommend as a workaround to use the ``Float`` type when interacting with triggers and grip buttons and apply your own threshold.

Author: No, thats taken care off automatically. So a controller that has a button trigger or button grip and thus has the input defined as `Bool`, if it is bound to an action of type `Float` the conversion is correct. The output to the game will be 0.0 if depressed and 1.0 if pressed.

The problem is with analogue inputs, for instance on the grip the trigger value input is a `Float`, if you bind it to a `Bool` action it has some really low threshold like 0.1 or something, so you have to barely touch your controller and it already returns you're gripping.

It's why we had major problems with the ball throwing game in our demo with the ball not letting go.

There are some improvements on the horizon here but for now it's better to use actions typed as `float` when using analogue buttons.

>   We recommend as a workaround to use the ``Float`` type when interacting with triggers and grip buttons and apply your own threshold.

Reviewer 3: I did see `float` degrade to `bool` but somehow missed that it also works the other way around. With that figured out, it now makes perfect sense. Sorry for the confusion!

This generic name is chosen because controllers don't cover the entire system. Currently there are also profiles for trackers, remotes and tracked pens. There are also provisions for devices such as treadmills, haptic vests and such even though those are not part of the specification yet.

.. warning::
  It is important to know that OpenXR has strict checking on supported devices. The core specification identifies a number of controllers and similar devices with their supported inputs and outputs, every XR runtime must accept these interaction profiles even if they aren't applicable. New devices are added through extensions and XR runtimes must specify which they support. XR runtimes that do not support a device added through extensions will not accept these profiles. XR runtimes that do not support added input or output types will often crash if supplied.

  As such Godot keeps meta data of all available devices, their inputs and outputs and which extension adds support for them. You can create interaction profiles for all devices you wish to support, Godot will filter out those not supported by the XR runtime the user is using.

>   This does mean that in order to support new devices, software changes are needed.

Reviewer 3: What software? I assume:
```suggestion
  This does mean that in order to support new devices, you might need to update to a more recent version of Godot.
```

It is however also important to note that the action map has been designed with this in mind. When new devices enter the market, or when your users use devices that you do not have access to, the action map system puts the onus with the XR runtime. It is the XR runtimes job to choose the best fitting interaction profile that has been specified and adapt it for the controller the user is using.

> How the XR runtime does this is left to the implementation of the runtime and there are thus vast differences between the runtimes.

Reviewer 3: ```suggestion
How the XR runtime does this is left to the implementation of the runtime and there are thus vast differences between the runtimes. Some runtimes might even permit users to edit the bindings themselves.
```

(You included this in a warning later on, but I feel this makes sense to mention here.)

A common approach for a runtime is to look for a matching interaction profile first, if this is not found check the most common profiles such as that of the "Touch controller" and do a conversion, and only if all else fails, check the generic "Simple controller", more on that one in a bit.

.. note::

>   There is an important conclusion to be made here. When a controller is found and the action map applied to it, the XR runtime will choose the most applicable interaction profile and enact the bindings. At this point it will communicate this choice to the game. We have to be careful that we do not assign a conclusion to this that can't be made.

Author: I join the club of non-native speakers, which is why I tent to be overly verbose :P

>   There is an important conclusion to be made here. When a controller is found and the action map applied to it, the XR runtime will choose the most applicable interaction profile and enact the bindings. At this point it will communicate this choice to the game. We have to be careful that we do not assign a conclusion to this that can't be made.

Reviewer 4: `to the game.` => `to the game/application.`

  For example, when the Touch controller profile is used any of the following scenarios could be true:

   * we could be using a Quest 1 controller,
   * we could be using a Quest 2 controller,
   * we could be using a Quest Pro controller but no Quest Pro profile was given or the XR runtime being used does not support the Quest Pro controller,

>    * it could be a completely different controller for which no profile was given but the XR runtime is using the touch bindings as a base.

Reviewer 2: ```suggestion
    * it could be a completely different controller for which no profile was given but the XR runtime is using the touch bindings as a base.
```

  At the moment none of the XR runtimes offer this functionality though SteamVR has an existing UI from OpenVRs action map system that is still accessible. This is actively being worked on however.

Our first controller binding
----------------------------

Lets setup our first controller binding and we'll do so for the Touch controller.

Press "Add profile" and find the touch controller and add it, if it is not in the list it may already have been added:

.. image:: img/xr_add_touch_controller.webp

Our UI now shows a panel for the left controller, and for the right controller. It shows all the possible inputs and outputs of the controller in this UI. We can use the ``+`` next to each entry to bind it to an action:

.. image:: img/xr_select_action.webp

Lets finish our configuration:

.. image:: img/xr_touch_completed.webp

Each action is bound the the given input or output for both controllers to indicate that we support the action on either controller. The exception is the movement action which is bound only to the right hand controller. It is likely that we would want to use the left hand thumbstick for a different purpose, say a teleport function.

> In developing your game however you have to account for the possibility that the user changes the binding and binds the movement to the left hand thumbstick. 

Reviewer 4: `game` => `game/application`

Also note that our shoot and grab boolean actions are linked to inputs of type ``float``. As mentioned before OpenXR will do conversions between the two, but do read the warning given on that subject earlier in this document.

The simple controller
---------------------

The "Simple controller" is a generic controller that OpenXR offers as a fallback. However as becomes painfully clear after we apply our mapping:

.. image:: img/xr_simple_controller.webp

The simple controller is far too simple and falls short for anything but the simplest of VR games.

This is why many XR runtimes only use it as a last resort and will attempt to use bindings from one of the more popular systems as a fallback first.

.. note::
  Due to the fallback not being very useful it is tempting to provide bindings for every controller supported by OpenXR. This is what the default action map does.

>   However it is the recommendation from the OpenXR Working Group that only bindings for controllers actually tested by the developer are setup. The XR runtimes are designed with this in mind and can do a better job at rebinding a provided binding, then a developer who takes a best stab at bindings without testing if this leads to a comfortable experience for the end user.

Reviewer 3: Does the WG have a recommendation regarding the simple controller? Wouldn't they want that one to be included if possible?

Author: Yes they want it included but they recognize that it falls far short for most games and is practically unusable. I can't say what they plan to do about it as that is still being debated, but it is on their radar.

Author: Also note, from experimentation and looking at the SDKs I know the following:
- SteamVR allows driver developers to nominate a primary fallback which by default is the Touch controller, seeing Touch is probably the most prevalent controller that's a pretty safe bet
- WMR has paths to nearly all major controllers, so if you have either a Vive Wand, Touch or Index controller profile, they'll have mappings to MR and G2 controllers.
- Meta I don't know for sure, but seeing you'll be testing on a Quest if you want to deploy to their store it's a given you'll be providing a Touch profile.
- PICO we know falls back to simple controller and that has been a pain
- Not sure what Monado does but pretty sure they do more intelligent stuff as well

I'm hesitant to put that into the docs though as these things can change rapidly.

Reviewer 3: Definitively reasonable to not go into detail here. I just wondered whether the simple controller would be useful to include for very primitive games (even though people generally don't test on the simple controller, for the lack of that controller being a real thing).

Author: It is, look at a game like Beat Saber, it could fully run with just specifying the simple controller. That is why I find it important to make people aware off its existence, but reality is that it just isn't good enough for most applications.

Author: I've expanded on this section a little, hopefully the intention is a bit clearer now.

    * Godot's input system mainly centers around button inputs, XR adds triggers, axis, poses and haptics (output) into the mix. This would greatly complicate the input system with features that won't work for normal controllers or contrast with the current approach. It was felt this would lead to confusion for the majority of Godot users.
    * Godot's input system works with raw input data that is parsed and triggers emitting actions, this input data is made available to the end user. OpenXR completely hides raw data and does all the parsing for us, we only get access to already parsed action data. This inconsistency is likely to lead to bugs when an unsuspecting user tries to use an XR device as a normal input device.
    * Godot's input system allows changes to what inputs are bound to actions in runtime, OpenXR does not.
    * Godot's input system is based on device ids which are meaningless in OpenXR.

>   This does mean that a game/application that mixes traditional inputs with XR controllers will have a separation. For most applications either one or the other is used and this is not seen as a problem. In the end, it's a limitation of the system.

Reviewer 5: What exactly is meant here by "a separation" does that mean a project at runtime can only use one or the other?

Author: It means that you'll be implementing two separate systems. You'll be handling keyboard/mouse/HOTAS input through Godots existing input system, and XR controllers through the action system. You can't create an interface where the user can optionally map an action to either traditional or XR inputs.

In practice this generally isn't a problem.

The majority of XR games only feature XR input as most now follow the traditional "Quest" like model of a headset with two XR controllers in hand.

Those that don't, like driving sims or flight sims, go the other way ignoring XR controllers in favor of racing wheels and HOTAS.

There are a very tiny number of use cases were people want to mix and match but I do want to underline here that the choice was made for us. OpenXR takes over input binding and parsing and has closed the door to easily supporting this scenario and that as I understand it, was a conscious decision.

The concept here is that you could have different sets that provide bindings in different scenarios. You could have a set ``Character control`` for when you're walking around, ``Vehicle control`` for when you're operating a vehicle, ``Menu`` for when a menu is open, etc.
Only the action set applicable to the current state of your game/application can then be enabled.

This is especially important if you wish to bind the same input on a controller to a different action. For instance:

  * in your ``Character control`` set you may have an action ``Jump``, 
  * in your ``Vehicle control`` set you may have an action ``Accelerate``, 
  * in your ``Menu`` set you may have an action ``Select``.

All are bound to the trigger on your controller.

OpenXR will only bind an input or output to a single action. If the same input or output is bound to multiple actions the one in the active action set with the highest priority will be the one updated/used. So in our above example it will thus be important that only one action set is active.

For your first XR game/application we highly recommend starting with just a single action set and not over-engineer things.

> For our walk through in this document we will thus create a single action set called ``my_first_action_set``, we do this by pressing the ``Add action set`` button:

Reviewer 6: ```suggestion
For our walkthrough in this document we will thus create a single action set called ``my_first_action_set``, we do this by pressing the ``Add action set`` button:
```

---------------------

The "Simple controller" is a generic controller that OpenXR offers as a fallback. However as becomes painfully clear after we apply our mapping:

.. image:: img/xr_simple_controller.webp

The simple controller is often far too simple and falls short for anything but the simplest of VR games/applications.

This is why many XR runtimes only use it as a last resort and will attempt to use bindings from one of the more popular systems as a fallback first.

.. note::
  Due to the simple controller likely not covering the needs of your game, it is tempting to provide bindings for every controller supported by OpenXR. The default action map seems to suggest this as a valid course of action however as mentioned before, the default action map was designed for ease of migration from Godot 3.

  It is the recommendation from the OpenXR Working Group that only bindings for controllers actually tested by the developer are setup. The XR runtimes are designed with this in mind and can do a better job at rebinding a provided binding, then a developer who takes a best stab at bindings without testing if this leads to a comfortable experience for the end user.

>   This is our advise as well, limit your action map to the interaction profiles for devices you have actually tested your game with. The Oculus Touch controller is widely used as a fallback controller by many runtimes, if you are able to test your game using a Meta Rift or Quest and add the profile for the Touch controller there is a high probability your game will work with other headsets.

Reviewer 6: ```suggestion
  This is our advice as well: limit your action map to the interaction profiles for devices you have actually tested your game with. The Oculus Touch controller is widely used as a fallback controller by many runtimes, if you are able to test your game using a Meta Rift or Quest and add the profile for the Touch controller there is a high probability your game will work with other headsets.
```
